plugins/pulp_rpm/plugins/db/models.py

```diff
@@ -1,93 +1,141 @@
-
-        return cls(**unit_key)
-
-    def clean_metadata(self):
-        """
-        Iterate through each key in the "metadata" dict, and if it starts with
-        a "_", delete it. This is to clean out mongo-specific and platform-specific
-        data. In the future, this will likely go away if we more strongly define
-        which fields each model will hold.
-        """
-        for key in self.metadata.keys():
-            if key.startswith('_'):
-                del self.metadata[key]
+    def clone(self):
+        """
+        Creates a new instance of an equivalent unit, with the unique ID missing. This is useful
+        for the PackageGroup and similar units, where there is a regular need to create a copy of
+        a unit, changing just one field in the unit key.
 
-    def __str__(self):
-        return '%s: %s' % (self.TYPE, '-'.join(getattr(self, name) for name in self.UNIT_KEY_NAMES))
+        This is a strange use case, so at this point it doesn't seem worth putting this method
+        into the platform.
+
+        deepcopy does not appear to work, presumably because of all the special handling that
+        mongoengine does in the metaclass. Suggestions are welcome for a better way to clone.
+
+        :return:    a new unit instance without a unique ID
+        :rtype:     Package
+        """
+        son_data = self.to_mongo()
+        son_data.pop('_id')
+        return self.__class__(**son_data)
+
+
+class NonMetadataPackage(Package):
+    """
+    An abstract model to be subclassed by packages which are not metadata
+
+    It is tempting to have checksum auto-calculate as a pre_save_signal(), but with checksum being
+    part of the unit_key it is needed when __init__() is called. The filename is not provided to
+    __init__() which makes doing it then impossible. For now, the checksum field is the
+    responsibility of the unit instantiator.
+
+    :ivar version: The version field of the package
+    :type version: mongoengine.Stringfield
+
+    :ivar release: The release field of the package
+    :type release: mongoengine.Stringfield
+
+    :ivar checksumtype: The checksum type of the package.
+    :type checksumtype: mongoengine.StringField
+
+    :ivar checksum: The checksum of the package.
+    :type checksum: mongoengine.StringField
+
+    :ivar version_sort_index: ???
+    :type version_sort_index: mongoengine.StringField
+
+    :ivar release_sort_index: ???
+    :type release_sort_index: mongoengine.StringField
+    """
+
+    version = mongoengine.StringField(required=True)
+    release = mongoengine.StringField(required=True)
+    checksumtype = mongoengine.StringField(required=True)
+    checksum = mongoengine.StringField(required=True)
+
+    # We generate these two
+    version_sort_index = mongoengine.StringField()
+    release_sort_index = mongoengine.StringField()
+
+    meta = {
+        'abstract': True,
+    }
+
+    def __init__(self, *args, **kwargs):
+        if 'checksumtype' in kwargs:
```

Comment:
I agree, an excellent suggestion. We would subclass StringField and call it `ChecksumTypeStringField` and add this the check to the validate() there right?

Also I propose that we keep the check here in `__init__` since it's part of the unit key in many units we want invalid values raised as early as possible. What do you think?